File: 数组/3sum/3sum.py
# 用map实现，遍历2个数，找-i-j是否在map中
# 时间复杂度 O(n^2)
# 空间复杂度 O(n)
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []
        res = set()
        nums.sort()
        for i,v in enumerate(nums[:-2]):
            # trick：相同的值只遍历1次
            if i >= 1 and v == nums[i-1]:
                continue
            d = {} # 查看是否包含第三个数
            for x in nums[i+1:]:
                if x not in d:
                    d[-v-x] = 1
                else:
                    res.add((v,-v-x,x))
        return list(res)
                

File: 数组/16. 3sum/3sum.cpp
/* 思路2：用2个指针，第一个数遍历nums[:-2]，j指向i+1，k指向末尾
 *       如果nums[i] + nums[j] + nums[k] < 0，则j需要向后移动，同时判断重复
 *       如果nums[i] + nums[j] + nums[k] > 0，则k需要向前移动，同时判断重复
 *       如果满足，添加结果，并j,k同时移动一步
 *       当j >= k,退出循环
 * 时间复杂度 O(n^2)
 * 空间复杂度 O(n)
 */
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if(nums.size() < 3) return vector<vector<int>>();
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for(int i = 0; i < nums.size()-2; i++) {
            if(i > 0 && nums[i] == nums[i-1]) continue;
            for(int j=i+1,k=nums.size()-1; j < k; ) {
                if(j > i+1 && nums[j] == nums[j-1]) {
                    j++;
                } else if(k < nums.size()-1 && nums[k] == nums[k+1]) {
                    k--;
                } else {
                    if(nums[i] + nums[j] + nums[k] > 0) k--;
                    else if(nums[i] + nums[j] + nums[k] < 0) j++;
                    else {
                        res.push_back({nums[i],nums[j],nums[k]});
                        j++;
                        k--;
                    }
                }
            }
        }
        return res;
    }
};